google-cloud-spanner/lib/google/cloud/spanner/client.rb

```diff
@@ -371,19 +375,35 @@ def query_options
         #     puts "User #{row[:id]} is #{row[:name]}"
         #   end
         #
+        # @example Query using tag for request query statistics collection.
+        #   require "google/cloud/spanner"
+        #
+        #   spanner = Google::Cloud::Spanner.new
+        #
+        #   db = spanner.client "my-instance", "my-database"
+        #
+        #   results = db.execute_query "SELECT * FROM users", tag: "Tag-1"
+        #
+        #   results.rows.each do |row|
+        #     puts "User #{row[:id]} is #{row[:name]}"
+        #   end
+        #
         def execute_query sql, params: nil, types: nil, single_use: nil,
-                          query_options: nil, call_options: nil
+                          query_options: nil, tag: nil,
```

Comment:
I wonder if the tag should not be in the `query_options` or `call_options` hashmap, wdyt? Or is this a private method used internally only?

Reply:
No, this is not a private method. In my view, we should not use `query_options` or `call_options`.`query_options` for optimization params and `call_options` is not for apis it is for GRPC client. We can use `tag` as `request_options` as per [proto file](https://github.com/googleapis/googleapis/blob/0dd48adfa0aa8f4d042813af248dcbeecc5f7102/google/spanner/v1/spanner.proto#L566) but client design specs "tag" is used. i.e ```roTx.read(table, keySet, columns, Tag.of("tag-2"));``` 

@thiagotnunes  Let me know your thoughts.

Reply:
I think it would be nice to define a `request_options`, because that will make the API more extensible (we have other features coming soon, for which we could piggy back on this new hashmap then).

For the client design spec, we have the `roTx.read(table, keySet, columns, Tag.of("tag-2"));`, but if you check the implementation of that, you can see that `Tag` is actually a flavour of `Option` as well (https://github.com/googleapis/java-spanner/pull/576/files#diff-992d480774d874d97423da09934b0395ef460130558e2fd6f7b1461ba32535c0R154-R166). So, if we have the `request_options` map here, we would not be deviating from the original design.

Reply:
@thiagotnunes Thanks for detailed information. Making changes according  to it.

Reply:
@thiagotnunes  So method will look like this. Please confirm.

```
 def execute_query sql, params: nil, types: nil, single_use: nil,
                                  query_options: nil, call_options: nil
                                  query_options: nil, request_options: nil,
```
Use:
```
client.execute_query("SELECT ...", request_options: { request_tag: "Tag-1" })
```

For transaction:
```
def transaction deadline: 120, request_options: nil, call_options: nil
```
Use
```
  client.transaction request_options: { transaction_tag: "Tag-1 } do 
  end
```

Reply:
I think this is a reasonable design. When you implement it, just make sure that in the documentation for the function, it is clear if the other type of tag is ignored. Something like "if transaction_tag is specified, it will be ignored".

Reply:
ok. @skuruppu Do we need to implement ignore tag logic or just documentation is sufficient?

Reply:
@jiren I think that the user should never have to decide whether the tag should be a `request_tag` or `transaction_tag`, but that should be done internally. We could change the example to:

```
client.transaction request_options: { tag: "Tag-1 } do 
end
```

```
client.execute_query("SELECT ...", request_options: { tag: "Tag-1" })
```

Wdyt?

Reply:
@thiagotnunes yes this will be helpful for users. I will start making changes for this.

Reply:
@thiagotnunes Updated as per above examples.